components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/LegacyJobOrchestrator.cs

```diff
@@ -756,13 +756,89 @@ IJsonSerializer jsonSerializer
             PublishedNodesEntryModel request,
             CancellationToken ct) {
             _logger.Information("{nameof} method triggered", nameof(UnpublishAllNodesAsync));
+            var sw = Stopwatch.StartNew();
             await _lockConfig.WaitAsync(ct).ConfigureAwait(false);
+            var response = new List<string>();
             try {
-                throw new MethodCallStatusException((int)HttpStatusCode.NotImplemented, "Not Implemented");
+
+                // Create HashSet of nodes to remove.
+                var nodesToRemoveSet = new HashSet<OpcNodeModel>(OpcNodeModelEx.Comparer);
+
+                // Perform first pass to determine if we can find all nodes to remove.
+                var matchingGroups = new List<PublishedNodesEntryModel>();
+                foreach (var entry in _publishedNodesEntries) {
+                    if (entry.HasSameGroup(request)) {
+                        // We may have several entries with the same DataSetGroup definition,
+                        // so we will remove nodes only if the whole DataSet definition matches.
+                        if (IsSameDataSet(entry, request)) {
+                            entry.OpcNodes.Select(n => nodesToRemoveSet.Add(n));
+                            entry.OpcNodes.Clear();
+                            matchingGroups.Add(entry);
+                        }
+                    }
+                }
+
+                // Report error if no matching endpoint was found.
+                if (matchingGroups.Count == 0) {
+                    throw new MethodCallStatusException((int)HttpStatusCode.NotFound,
+                        $"Endpoint not found: {request.EndpointUrl}");
+                }
+
+                // Report error if there were entries that did not have any nodes
+                if (nodesToRemoveSet.Count == 0) {
+                    request.OpcNodes = nodesToRemoveSet.ToList();
```

Comment:
The if condition here already makes sure that `nodesToRemoveSet` is going to be an empty list, so this conversion to list doesn't add any value. Also, why do we need to change the request here just before throwing an exception ? I think that we do not need to have `nodesToRemoveSet` which is a `HashSet` at all. A simple bool flag would be sufficient to check if we have found any nodes or not.